### PR TITLE
[AutoDiff] Add JVP derivatives for `fmod` and `remainder`

### PR DESCRIPTION
**Pull request by philipturner**
Resolves TF-1108.

The original link got lost when the TensorFlow project was deleted from JIRA. I can't find it by looking up keywords, although maybe I'm not looking hard enough. There is a test failure about ULP error being unsatisfactory. Here's a snippet of the error log when running `tgmath_derivatives.swift.gyb`:
```
--
Exit Code: 250

Command Output (stdout):
--
[ RUN      ] TGMath.derivative_Float
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> -2.0 != -1.999974250793457 as Float
stdout>>>   216.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> -2.0 != -1.999974250793457 as Float
stdout>>>   216.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> 3.0 != 3.0000686645507812 as Float
stdout>>>   288.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> 3.0 != 3.0000686645507812 as Float
stdout>>>   288.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> 3.0 != 3.0000686645507812 as Float
stdout>>>   288.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> -1.0 != -0.9999752044677734 as Float
stdout>>>   416.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
...
stdout>>> expected: true
stdout>>> 0.0 != 2.384185791015625e-05 as Float
stdout>>>   13107200.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
stdout>>> expected: true
stdout>>> 0.0 != 2.384185791015625e-05 as Float
stdout>>>   13107200.0-ulp error exceeds 192.0-ulp tolerance.
stdout>>> check failed at /Users/philipturner/Documents/Swift-Compiler/swift-project/swift/test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb, line 74
...
[     FAIL ] TGMath.derivative_Float
[ RUN      ] TGMath.gradient_Float
[       OK ] TGMath.gradient_Float
TGMath: Some tests failed, aborting
UXPASS: []
FAIL: ["derivative_Float"]
SKIP: []
```

At one point, it reached 13 million ULP, which is a lot larger than the tolerance of 192. I have little experience with JVP derivatives, so I don't know if I'm doing them wrong. The last time the relevant code was modified was [2 years ago](https://github.com/philipturner/swift/commit/f07ae8963614d65f8783f6b26560621665636a98), so the ULP failure probably hasn't been investigated thoroughly.

Can this PR be reported to the person who filed TF-1108 (if known)? Should I increase the tolerance on single-precision for the offending functions?

### CODE REVIEW

**Comment by philipturner** [comment]
CC: @rxwei

**Comment by philipturner** [comment]
This will create a merge conflict with the solution for double-precision differentiation, https://github.com/apple/swift/pull/40991. Richard Wei, sorry for pinging you so many times in a 24-hour period. Should the one [here](https://github.com/tensorflow/swift-apis/issues/1185#issuecomment-1019766631) have gone to Brad Larson instead?

**Comment by rxwei** [comment]
> Richard Wei, sorry for pinging you so many times in a 24-hour period. Should the one [here](https://github.com/tensorflow/swift-apis/issues/1185#issuecomment-1019766631) have gone to Brad Larson instead?

I'm not going to comment on that issue since it's not directly related to the Swift Open Source Project.  Yes, I think @BradLarson is a good person to talk to about that project.

**Comment by stephentyrone** [comment]
These functions are not (continuously) differentiable, only piecewise (they're not even continuous). What's the policy for handling such functions with autodiff; should they have derivatives at all? What is supposed to happen at the discontinuities?

**Comment by philipturner** [comment]
@stephentyrone discontinuous functions are differentable at the points where they're continuous. If you reach a discontinuity, it's like you approach it from one side and take the limit of the derivative. We even have differentiation of control flow, which is discontinuous by my understanding.

**Comment by stephentyrone** [comment]
Yes, I understand the mathematics perfectly. I am asking what the policy for the software package is. Is the expected behavior documented somewhere?

**Comment by philipturner** [comment]
If you want an in-depth explanation, I suggest checking out the following:
- https://github.com/apple/swift/blob/main/stdlib/public/Differentiation/DifferentialOperators.swift
- https://github.com/apple/swift/blob/main/docs/DifferentiableProgramming.md
- https://github.com/apple/swift/blob/main/docs/DifferentiableProgrammingImplementation.md

**Comment by stephentyrone** [comment]
I don't want an in-depth explanation, I want an answer to the question "has the autodiff project defined a policy for what the derivative should be at a step discontinuity?"

**Comment by rxwei** [comment]
@philipturner I'm afraid you are missing @stephentyrone's point.  The differentiation of control flow is very well understood, but it is irrelevant here, because control flow differentiation the semantics of the language, while Steve is asking about Swift's math library.

I'm of the opinion that we should _not_ define piecewise derivatives for math library.  I'm not sure why the VJPs of `fmod` and `remainder` slipped in and what was the rationale.  I think someone implemented it after TensorFlow/PyTorch.

**Comment by philipturner** [comment]
@rxwei removing them would make it easier in that I can skip my draft pull request for JVP fmod/remainder. But, wouldn't that leave an empty hole in that 90% of the official tgmath functions have derivatives while 10% don't?

**Comment by stephentyrone** [comment]
I'm not really opposed to defining derivatives for these, but there should be an actual policy for what happens at the jumps. If we don't have one yet, looking at what Julia does might be instructive. Either way, it should be clearly documented.

**Comment by philipturner** [comment]
@stephentyrone that sounds like a good idea to me. I think it's best if @rxwei gives his say first, but I'm willing to spend some time polishing up the documentation for something like that.

**Comment by rxwei** [comment]
If you can look into other implementations, that'd be great.  I'm curious to know what Julia (or PyTorch or TensorFlow) does.

**Comment by philipturner** [comment]
@stephentyrone since I’m going to activate this pull request soon, could you move your conversation to another thread? I am open to helping you contribute to the documentation.

**Comment by rxwei** [comment]
I'm not sure that makes any sense. @stephentyrone is reviewing your PR with perfectly on-topic comments, and you need to address his feedback as part of this PR.

**Comment by philipturner** [comment]
@stephentyrone sorry for misinterpreting the conversation. I was thinking that you were asking to add text to documentation on AutoDiff somewhere. You were actually referring to the fact that jump discontinuities in this pull request are concerning. I won’t open this PR yet, now that I have a full understanding of this conversation.

One big area of concern I mentioned above is that there’s an unusual error rate with the functions I added. Could you lend me your expertise or refer to other implementations (e.g. Julia, PyTorch) about why this might happen?

**Comment by philipturner** [comment]
Perhaps we should strive to match the set of elementary functions in Swift Numerics?

**Comment by philipturner** [comment]
> a. land a PR that simply removes the derivatives of all the remainder functions.

I noticed that the derivative of `abs(_:)` is missing from all 3 *Differentiation.gyb files (FloatingPoint, SIMD, Tgmath). That is a jump discontinuity and would have to be added if we proceed as we are doing now, because it's an essential function for numeric values. Plus, `ElementaryFunctions` from Swift Numerics doesn't include `abs(_:)` or any functions with jump discontinuities.

I think (a) is the safest path forward. I have to research what clients use the stdlib's built-in differentiation for jump-discontinuous functions though.

**Comment by stephentyrone** [comment]
`abs()` is less problematic; there's no derivative at the origin, but the function itself is continuous. A derivative of either zero or NaN could be defended at the point.

`cosh`'s derivative is perfectly well-behaved; perhaps you mean `acosh`? In which case, the upper branch is principal, and so the derivative at 1 must be +infinity.

**Comment by philipturner** [comment]
> cosh's derivative is perfectly well-behaved; perhaps you mean acosh? In which case, the upper branch is principal, and so the derivative at 1 must be +infinity.

I didn't really know what I was talking about when I said `cosh(_:)`. I just knew that some hyperbolic or trig functions like `tan(_:)` might have wierd asymptotes.

> `abs()` is less problematic; there's no derivative at the origin, but the function itself is continuous.

`abs` falls into the same boat as `fmod` and `remainder`, as it has two conflicting derivatives coming in from each side. An optimized derivative implementation would consider what's at zero to be either +1 or -1, as explicitly checking `== 0` is an extra clock cycle. This is similar to the behavior of `FloatingPoint.minimum` derivatives, which have to choose between one of the inputs and can't compromise.

I propose that we examine the left and right hand derivative limits at a C1 jump discontinuity. If they agree, then use them as the derivative. Otherwise, make it `nan`. Thus, the derivative of `abs` at 0 would be `nan`. This would cost an extra clock cycle(s) and we may have to examine functions that previously worked, to ensure this policy is rigorously enforced.

With floating-point math, it's good practice to check for `nan`s in edge cases. I did that frequently when accelerating some math code for [AR MultiPendulum](https://github.com/philipturner/ar-multipendulum) on the GPU. If users don't like that, they can always override the stdlib with a custom derivative.

In summary, if we go for (b), there are two options:
1. Let the derivative at a C1 jump discontinuity take either side. In the fastest implementation, it might take either left or right, with no consistency.
2. Make special case derivatives for C1 jump discontinuities. This could cost extra clock cycles and be difficult to enforce. Plus, we could unexpectedly propagate `nan` into people's code.

**Comment by philipturner** [comment]
SR-15794

**Comment by philipturner** [comment]
@rxwei the reason these tests are failing is a fundamental flaw in the testing algorithm. The produced derivatives are perfect. For example, the JVP derivative of `fmod(-10, -3)` produces `-2.0`. But, `computeDividedDifference` produces a number `-1.999974250793457` to compare it against. The fake `-1.999974250793457` number is what's > 192.0 ulp off, not the derivative. 

We need to either fundamentally change this test mechanism or disable it for JVP derivative of `fmod(Float, Float)`.

**Comment by stephentyrone** [comment]
The pointwise derivative at the corner of a C0 function doesn't matter; whatever value is assigned there, the integral of the derivative will be equal to the original function. But when the function fails to be C0 due to a step discontinuity, there is no value that can be assigned to a derivative _function_ with this property (the notion of a [distributional derivative](https://en.wikipedia.org/wiki/Distribution_(mathematics)) can be used to patch this up, but that's is not something that autodiff is currently equipped to work with).

This is why I'm not concerned about `abs`. Whatever value is assigned there, the calculus "works". This is not true for `fmod` and similar functions, which is why I think we should simply not define derivatives for them at this time.

**Comment by philipturner** [comment]
> (likewise for tangent, except that you can never hit the jumps exactly in the case of tangent, so it does not matter at all).

I think you confused jumps with asymptotes there. You could hit that by passing in angle = pi / 2. Also, the derivative for the reciprocal function could be hit by setting x = 0. That's an asymptote. For example:
```swift
import _Differentiation

print(gradient(at: Float(0), of: { Float(1) / $0 }))
```
Output:
```
-inf
```

**Comment by stephentyrone** [comment]
> So you're suggesting we base Swift's derivatives based on the theory of what happens if we take the integral over it, rather than if the two directional limits of f' agree?

I'm saying that if you solve the differential equation f(0) = initial value, f'(x) = automatically-generated-derivative-of-f(x), do not get any indication of an error, and do not recover an approximation of the function f, that's a problem that limits the usefulness of autodiff and introduces subtle bugs. So functions with step discontinuities probably shouldn't be differentiable in the system, or should at least require special opt-in.

**Comment by philipturner** [comment]
The reason I related control flow to this a while ago is because you can synthesize `@differentiable(reverse)` step-discontinuous functions and AutoDiff makes derivatives for them. For example:

```swift
@differentiable(reverse)
func myStep(_ x: Float) -> Float {
  var output = 0
  if (x < 0) {
    output = -1
  } else {
    output = x
  }
  output *= 2
  return output
}
````
That's a step-discontinuous function, and the compiler doesn't crash while compiling it (if it does, that's a bug). However, the user specified that there's a certain behavior in the range `-infinity < x < 0` and in `0 <= x < infinity`. So there's no discontinuity because `x = 0` is included in the rightward interval. For the remainder functions, there's also a defined behavior where at the discontinuity, it is always included in the rightward interval (or maybe always left). Thus, manual derivatives for `remainder` would mirror what the compiler just did for `myStep`.

**Comment by stephentyrone** [comment]
I don't think you understand my objection; I am not saying that you cannot define a "derivative". I am saying that the derivative that you define doesn't have some of the properties that make a derivative useful for computation.

This is not a mathematical question. This is a _policy_ question that needs to be resolved before this patch can land: the project needs to decide and document how best to handle such functions.

**Comment by philipturner** [comment]
> I don't think you understand my objection

I understand your objection. We are figuring out how to set bizarre derivatives to something meaningful and reliable. I was showing that there is a precedent for how these kinds of derivatives should be defined. According to your response, the derivatives that Swift automatically synthesizes for `myStep` are not useful.

> This is not a mathematical question. This is a _policy_ question

Modifying a manual derivative to deviate from the `myStep` behavior could make it "useful", but inconsistent with automatic derivatives. In other words, I'm trying to make a policy that uniformly applies to automatic derivatives and derivatives in the standard library. 

Furthermore, such a policy would be easiest to implement and enforce. We could just make a custom function that mirrors the behavior of, say, `fmod`:

```swift
// Citing the Metal Shading Language Specification (10/05/2021, page 144)
@differentiable(reverse)
func fmodReplica(_ x: Float, _ y: Float) -> Float {
  return x - y * truncate(x / y)
}

for x in lowerX...upperX {
  for y in lowerY...upperY {
     assert(fmod(x, y) == fmodReplica(x, y))
  }
}

let toTestAgainst = pullback(at: x, y, of: fmodReplica)
```
Let the compiler synthesize the derivative, compare it to ours. Modify our test suite to validate against that, and we have a policy. If necessary, we can peek into the source code of Darwin and translate from C to Swift.

**Comment by philipturner** [comment]
Even better, do this:
```swift
let fmodTranslated = ... // fmod function translated from C to Swift

@derivative(of: fmod)
func _vjpFmod(_ x: Float, _ y: Float) -> (
  value: Float, pullback: (Float) -> (Float, Float)
) {
  return (fmod(x, y), pullback(at: x, y, of: fmodTranslated))
}
```

**Comment by ProfFan** [comment]
I agree with @stephentyrone . Discontinuity is a serious issue and should be thoroughly investigated. I would suggest first looking at JAX, since they have similar goals as S4TF.

**Comment by philipturner** [comment]
Yes, but what if the system library were written in Swift instead of C? Either way, we're still differentiating the same code, with the same output for the same input. But because it's in C, we override autodiff's synthesized derivatives with something that doesn't match? I think that Swift AutoDiff is a better precedent and guideline to follow than JAX.

**Comment by philipturner** [comment]
With your knowledge, could you construct a Swift manual derivative as if those assembly instructions were being processed by AutoDiff?

**Comment by philipturner** [comment]
> With your knowledge, could you construct a Swift manual derivative as if those assembly instructions were being processed by AutoDiff?

@stephentyrone any update on this? If it's too much to ask of you, then I could decompile the assembly myself.
